### PR TITLE
Feature/consensuscontext create

### PR DESCRIPTION
**Pull request by IdoZilberberg**

### CODE REVIEW

File: services/consensusalgo/leanhelix/algorithm.go [hunk 1]
```diff
@@ -22,33 +21,34 @@ func (s *service) leaderProposeNextBlockIfNeeded() error {
...
-			}).Build(),
-			TransactionReceipts: nil,
-			ContractStateDiffs:  nil,
-			BlockProof:          (&protocol.ResultsBlockProofBuilder{}).Build(),
-		},
+		TransactionsBlock: txBlock,
+		ResultsBlock:      rxBlock,
 	}
 
 	s.blocksForRoundsMutex.Lock()
```
Comment (electricmonk) [inline services/consensusalgo/leanhelix/algorithm.go]: Not part of this PR but I'd move this line and the following two to a local method `saveProposedBlock(atHeight primitives.BlockHeight)` that encapsulates the map and mutex logic

File: test/acceptance/consensus.go [hunk 2]
```diff
@@ -13,9 +12,9 @@ import (
 var _ = Describe("a leader node", func() {
 
 	It("must get validations by all nodes to commit a transaction", func(done Done) {
-		test.WithContext(func(ctx context.Context) {
-			// leader is nodeIndex 0, validator is nodeIndex 1
-			network := harness.NewTestNetwork(ctx, 2)
+		consensusAlgos := []consensus.ConsensusAlgoType{consensus.CONSENSUS_ALGO_TYPE_LEAN_HELIX}
```
Comment (electricmonk) [inline test/acceptance/consensus.go]: maybe a bit differently:
either extract the logic of creating the algos slice to a function that takes varargs of type `CONSENSUS_ALGO_TYPE`, so that the call would look like:

```golang
harness.WithNetwork(2, withAlgos(consensus.CONSENSUS_ALGO_TYPE_LEAN_HELIX), func(ctx context.Context, network harness.AcceptanceTestNetwork) {
}
```

or maybe create a network builder, like we had in the alpha:
```golang
harness.ANetwork()
  .WithNodes(2)
  .WithAlgos(consensus.CONSENSUS_ALGO_TYPE_LEAN_HELIX)
  .Run(func( network harness.AcceptanceTestNetwork) {
...
})

also, unless I'm missing something, the `context.Context` argument is being unused

File: services/consensusalgo/benchmarkconsensus/leader.go [hunk 3]
```diff
@@ -93,25 +94,31 @@ func (s *service) leaderGenerateNewProposedBlockUnderMutex() (*protocol.BlockPai
 
 	// get tx
 	txOutput, err := s.consensusContext.RequestNewTransactionsBlock(&services.RequestNewTransactionsBlockInput{
-		BlockHeight: s.lastCommittedBlockHeight() + 1,
+		BlockHeight:   s.lastCommittedBlockHeight() + 1,
+		PrevBlockHash: crypto.CalcTransactionsBlockHash(s.lastCommittedBlock.TransactionsBlock),
```
[the remark this answers is not in the document]
Comment (talkol) [reply services/consensusalgo/benchmarkconsensus/leader.go]: No, only the hash of prev block. The hash of current block header is not saved inside the block

File: services/consensusalgo/benchmarkconsensus/leader.go [hunk 4]
```diff
@@ -93,25 +94,31 @@ func (s *service) leaderGenerateNewProposedBlockUnderMutex() (*protocol.BlockPai
...
 	})
 	if err != nil {
 		return nil, err
 	}
+	if txOutput == nil || txOutput.TransactionsBlock == nil {
+		panic("invalid response consensusContext.RequestNewTransactionsBlock: missing TransactionsBlock")
```
[the remark this answers is not in the document]
Comment (talkol) [reply services/consensusalgo/benchmarkconsensus/leader.go]: These are dev only. They can only be nil if our internal code forgets to pass these arguments. We should remove these panics before going to production, right now they help with debugging because the code is missing in many places

File: services/consensusalgo/leanhelix/algorithm.go [hunk 5]
```diff
@@ -22,33 +21,34 @@ func (s *service) leaderProposeNextBlockIfNeeded() error {
 		return nil
 	}
 
-	proposedTransactions, err := s.transactionPool.GetTransactionsForOrdering(&services.GetTransactionsForOrderingInput{
-		MaxNumberOfTransactions: 1,
+	txOutput, err := s.consensusContext.RequestNewTransactionsBlock(&services.RequestNewTransactionsBlockInput{
+		BlockHeight:             s.lastCommittedBlockHeight + 1,
+		MaxBlockSizeKb:          0,
```
[the remark this answers is not in the document]
Comment (talkol) [reply services/consensusalgo/leanhelix/algorithm.go]: It's a smell left there on purpose so we remember to fill in these fields. They can't stay empty

File: services/consensusalgo/leanhelix/algorithm.go [hunk 6]
```diff
@@ -22,33 +21,34 @@ func (s *service) leaderProposeNextBlockIfNeeded() error {
...
 	if err != nil {
 		return err
 	}
 
+	txBlock := txOutput.TransactionsBlock
+	txBlock.BlockProof = (&protocol.TransactionsBlockProofBuilder{}).Build()
+
+	rxOutput, err := s.consensusContext.RequestNewResultsBlock(&services.RequestNewResultsBlockInput{
+		BlockHeight:       s.lastCommittedBlockHeight + 1,
+		PrevBlockHash:     nil,
```
[the remark this answers is not in the document]
Comment (talkol) [reply services/consensusalgo/leanhelix/algorithm.go]: It's also a smell left on purpose so we remember to fill in this field. Right now we just refactored the minimum to keep the acceptance working in lean helix and this wasn't required to make it pass.

The reason it's not needed is that lean helix today doesn't verify the prev block hash because it currently has a very partial implementation

File: services/consensusalgo/leanhelix/algorithm.go (hunk 1 above)
Comment (talkol) [reply services/consensusalgo/leanhelix/algorithm.go]: I'm not sure how much we should invest in improving lean helix because it's a temporary implementation.. it will be re-written completely by Gill's library. The lean helix basically transformed to benchmark consensus so let's invest there

File: test/acceptance/consensus.go (hunk 2 above)
Comment (talkol) [reply test/acceptance/consensus.go]: Fixed in https://github.com/orbs-network/orbs-network-go/pull/63
